prog_test.go

```diff
@@ -237,6 +237,28 @@ func TestProgramUnpin(t *testing.T) {
 	}
 }
 
+func TestProgramLoadPinnedWithFlags(t *testing.T) {
```

Comment:
It would be good to add one more test case that checks a legit case. Intercept calls to `BPFObjGet`, and verify that the flags bits are set correctly. One way is to introduce dependency injection by extracting the BPF* syscall wrappers to an interface.

Reply:
Yeah, I wanted to write such a test originally but I thought that it would require quite a bit of reworking the whole test suite. Writing down my rationale made me realise that I was wrong. I'll modify this to use BPF_F_RDONLY and then check that the loaded map can't be modified via Put. Thanks!

Other than that, I think we already know that the flag bits are set correctly:

1. Most tests don't set flags and therefore use 0 implicitly. We know this works.
2. Since the syscall with bogus flags fails we know that having non-zero flags is passed to the kernel.

Adding dependency injection might even weaken the second point, since there could be a bug in the syscall wrapper that we then miss.

Reply:
So this worked well for maps, not so well for programs and links. Programs are already kind of immutable, so I don't know how to test the effect of BPF_F_RDONLY. For a link the kernel seems to ignore BPF_F_RDONLY, which seems dodgy to me. I've keep program and link tests as they are right now.

Reply:
Flags are ignored for links and programs: https://elixir.bootlin.com/linux/latest/source/kernel/bpf/inode.c#L545

Reply:
I sent a mail to bpf@ https://lore.kernel.org/bpf/CACAyw99e288cPoBuxTjt17YfMy8AHT72AmS1W83EexxvWKaP3w@mail.gmail.com/T/#u

Reply:
If a unit test fails, how can you tell if there is a bug in the library or kernel?

Reply:
I don't understand what you're asking, sorry. Can you rephrase it?

Reply:
If the unit test that checks a legit case fails, how do we know if it's because the correct flags weren't set, or there was a bug in the kernel?

Reply:
Ah ok, you're saying that right now we don't get separate errors saying "the library is wrong!" and "the kernel is wrong!", which you would find useful. That's true! I think it's not enough of a benefit to start abstracting the syscalls away however. It implies writing tests for the raw syscalls that can tell us if the kernel is wrong, which would pretty much duplicate the tests for the higher level API (in an ideal world). It's better to assume that the error is in the library until proven otherwise.

Reply:
Stubbing/mocking out syscalls sounds like a lot of work for arguably little gain. Bugs in the kernel are usually detected by testing against a battery of kernels (like this lib does), and we all know the kernel "does not break userland, period". :slightly_smiling_face:

When it comes to kernel interfaces, I've always found more value in integration testing rather than unit testing bits in/out, since a stub/mock would have to emulate all of the kernel's corner cases and idiosyncracies, too. I haven't found a noteworthy iteration speed difference, either. Well-written integration test suites can/should complete in seconds.

Reply:
Ok, that's fair. @lmb Coming back to the original comment, what's the final testing plan to test legit cases?